internal/controllers/gateway/gateway_controller.go

```diff
@@ -0,0 +1,594 @@
+	for _, clusterIP := range svc.Spec.ClusterIPs {
+		for _, port := range svc.Spec.Ports {
+			addresses = append(addresses, gatewayv1alpha2.GatewayAddress{
+				Type:  &gatewayIPAddrType,
+				Value: clusterIP,
+			})
+			listeners = append(listeners, gatewayv1alpha2.Listener{
+				Name:          gatewayv1alpha2.SectionName(port.Name),
+				Protocol:      gatewayv1alpha2.ProtocolType(port.Protocol),
+				Port:          gatewayv1alpha2.PortNumber(port.Port),
+				AllowedRoutes: &r.allowedRoutes,
+			})
+		}
+	}
+
+	// for LoadBalancer service types we'll also capture the LB IP or Host
+	if svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
+		// if the loadbalancer IPs/Hosts haven't been provisioned yet
+		// the service is not ready and we'll need to wait.
+		if len(svc.Status.LoadBalancer.Ingress) < 1 {
+			err = fmt.Errorf("gateway service %s/%s is not yet ready (no LoadBalancer IP/Host provisioned)", svc.Namespace, svc.Name)
+			return
+		}
+
+		// otherwise gather any IPs or Hosts provisioned for the LoadBalancer
+		// and record them as Gateway Addresses.
+		for _, ingress := range svc.Status.LoadBalancer.Ingress {
+			if ingress.IP != "" {
+				addresses = append(addresses, gatewayv1alpha2.GatewayAddress{
+					Type:  &gatewayIPAddrType,
+					Value: ingress.IP,
+				})
+			}
+			if ingress.Hostname != "" {
+				addresses = append(addresses, gatewayv1alpha2.GatewayAddress{
+					Type:  &gatewayHostAddrType,
+					Value: ingress.Hostname,
+				})
+			}
+		}
+	}
+
+	return
+}
+
+// updateAddressesAndListenersSpec updates a pre-provisioned gateway's spec with new addresses and listeners.
+// If the addresses and listeners provided are the same as what exists, the update is skipped.
+func (r *GatewayReconciler) updateAddressesAndListenersSpec(ctx context.Context, obj *gatewayv1alpha2.Gateway, addresses []gatewayv1alpha2.GatewayAddress, listeners []gatewayv1alpha2.Listener) (bool, error) {
+	// if the existing object is already configured with the addresses and listeners that were
+	// identified for it, then no change needs to be made.
+	if reflect.DeepEqual(obj.Spec.Addresses, addresses) && reflect.DeepEqual(obj.Spec.Listeners, listeners) {
+		return false, nil
+	}
+
+	// override any existing addresses and listeners with the new ones provided and
+	// update the status of the object in the API.
+	obj.Spec.Listeners = listeners
+	obj.Spec.Addresses = addresses
+	return true, r.Update(ctx, obj)
+}
+
+// updateAddressesAndListenersStatus updates a pre-provisioned gateway's status with new addresses and listeners.
+// If the addresses and listeners provided are the same as what exists, it is assumed that reconciliation is complete and a Ready condition is posted.
+func (r *GatewayReconciler) updateAddressesAndListenersStatus(
+	ctx context.Context,
+	obj *gatewayv1alpha2.Gateway,
+) (bool, error) {
+	if !readyConditionExistsForObservedGeneration(obj) {
+		obj.Status.Listeners = convertListenerToListenerStatus(obj)
```

Comment:
I'd be uncomfortable merging this PR without an integration test that validates that the status fields get populated appropriately.

Reply:
This test already existed to do that: https://github.com/Kong/kubernetes-ingress-controller/pull/1986/files#diff-1cde2520ec6baa65f40a27801dbee1c6d60045bfdc2dfa1c36aec9785ee04211R101